R/tree-map-alignment.R

```diff
@@ -455,16 +460,344 @@ find_best_shift_icp = function(pred, obs) {
+  xy_mat1 = pred[c("x", "y")]
+  xy_mat2 = obs[c("x", "y")]
+
+  # Centering the coordinates before matching to improve numerical stability
+  xy_mat1_mean = colMeans(xy_mat1[c("x", "y")])
+  xy_mat2_mean = colMeans(xy_mat2[c("x", "y")])
+  # The double transpose seems to be the easiest way to broadcast subtraction
+  # given the recycling rules
+  xy_mat1 = t(t(xy_mat1) - xy_mat1_mean)
+  xy_mat2 = t(t(xy_mat2) - xy_mat2_mean)
+
+  # Number of objects detected from each point cloud
+  N_objects_vect = c(nrow(xy_mat1), nrow(xy_mat2))
+
+  if (N_objects_vect[2] < k) {
+    k = N_objects_vect[2]
+    msg = paste0(
+      "Number of objects in dataset 2 (and thus the potential number of matches) ",
+      "is smaller than given parameter k. k was set to ",
+      N_objects_vect[2]
+    )
+    # TODO make this a warning
+    print(msg)
+  }
+
+  # First, we construct the feature descriptor for each object in each of
+  # the point clouds.
+  feature_info1 = compute_feature_descriptors(xy_mat1, R_local)
+  feature_info2 = compute_feature_descriptors(xy_mat2, R_local)
+  # Extract out the components of the result
+  feat_desc_mat1 = feature_info1[[1]]
+  char_thetas1 = feature_info1[[2]]
+  feat_desc_mat2 = feature_info2[[1]]
+  char_thetas2 = feature_info2[[2]]
+
+  # Compute a matrix of pairwise distances between feature
+  # descriptors of the two point clouds -> rows correspond to objects detected
+  # from the point cloud 2 and columns correspond to objects detected from
+  # the point cloud 1.
+  feat_desc_pdist = cdist(feat_desc_mat2, feat_desc_mat1)
+
+  # Find the index per row with the lowest value
+  nn_indices = apply(feat_desc_pdist, 1, which.min)
+  lowest_dists = feat_desc_pdist[cbind(1:N_objects_vect[2], nn_indices)]
+
+  # Now find the second nearest index for each point in the first point cloud
+  # Mask out the first highest value.
+  # Note this must be done after extracting the distances
+  feat_desc_pdist[cbind(1:N_objects_vect[2], nn_indices)] = NaN
+  # Find the closest remaining value
+  second_nearest_inds = apply(feat_desc_pdist, 1, which.min)
+  second_lowest_dists = feat_desc_pdist[
+    cbind(1:N_objects_vect[2], second_nearest_inds)
+  ]
+
+  # Sorting the nearest neighbor distance ratios so that we can rank the
+  # tentative matches from most reliable to least reliable
+  NNDR_vect = lowest_dists / second_lowest_dists
+  sorted_indices = sort(NNDR_vect, index.return = TRUE, decreasing = FALSE)$ix
+
+  # Then, we consider the k most promising tentative matches
+  max_n_of_matches = 0 # maximum number of matching objects
+  ## Initializing the parameters for the transformation of the best match.
+  t_best = c(0, 0)
+  theta_best = 0
+  idx_matches1_best = c() # ndices of matching objects in point cloud 1
+  idx_matches2_best = c() # indices of matching objects in point cloud 2
+
+  if (length(sorted_indices) < k) {
+    k = length(sorted_indices)
+    print(paste("Not enough valid correspondences, setting k to ", k))
```

Comment:
Nice, could alternatively be a `warning`, or you could use `cat` for the combination of `print` and `paste`.